### PR TITLE
feat(node): txn prioritization based on gas parameters 

### LINKED ISSUE #1183: Enforce new base fee in mempool after gas market base fee update
With the introduction of gas market, block gas limit and base fee rotation were introduced. However, there is a missing piece when the base fee is updated, it should be enforced in the mempool for message selections: i.e. message in mempool below the new base fee, should be deprioritised for message selection.

### PR DESCRIPTION
**Pull request by cryptoAtwill**
This is a follow up to gas market feature and closes #1183.

The implementation follows the method 2 described. A high-level review is as follows:
- Check txn loads the past 5 blocks' base fee and constructs a priority calculator.
- Finds the lowest base fee in the past 5 blocks.
- The txn priority is calculated as outlined in https://github.com/consensus-shipyard/ipc/blob/17796a0f156121606186119aea59bb29a17e26f8/fendermint/vm/interpreter/src/fvm/state/priority.rs#L47

### CODE REVIEW

File: fendermint/app/src/app.rs [hunk 1]
```diff
@@ -561,22 +595,7 @@ where
...
-
-                FvmExecState::new(
-                    ReadOnlyBlockstore::new(db),
-                    self.multi_engine.as_ref(),
-                    state.block_height.try_into()?,
-                    state.state_params,
-                )
-                .context("error creating check state")?
-            }
+            None => self.create_check_state()?,
```
Comment (cryptoAtwill) [inline fendermint/app/src/app.rs:598]: We can avoid loading the 5 blocks state params for `check_tx`, to do this, we must change the `commit` to hold some historical `check_state` instead of purging everything. Not sure if we should do in this PR.

File: fendermint/vm/interpreter/src/fvm/state/priority.rs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+        if msg.gas_fee_cap >= base_fee {
+            let i = msg.gas_fee_cap.clone() - base_fee + msg.gas_premium.clone();
+            i.atto()
+                .min(&BigInt::from(i64::MAX))
+                .to_i64()
+                .expect("clipped to i64 max")
+        } else {
+            0
+        }
```
Comment (raulk) [inline fendermint/vm/interpreter/src/fvm/state/priority.rs:51]: 1. These kinds of conditionals are easier to follow when written as an early return. If the gas_fee_cap is below the base fee, return 0.
2. Are we sure 0 is the right value here? Why not return i64::MIN if the message is non-includable? Right now messages with gas_fee_cap == base_fee && gas_premium == 0 would rank exactly the same as non-includable messages. Seems wrong?
3. Adding the gas premium to the delta between the gas_fee_cap and the base fee is incorrect. The gas_fee_cap _includes_ the payable base fee **and** the gas premium. When the message is includable, our priority should be proportional to the **effective gas premium**. See how the [effective gas premium](https://github.com/filecoin-project/ref-fvm/blob/a8f9be18e76fcbe5d4b326d8c213eedb58121996/fvm/src/executor/default.rs#L91) is calculated in ref-fvm; copy the logic here.
4. i64::MAX atto is equal to 9.223372036854775807 base units. We are probably safe in clipping the effective premium to i64::MAX, since nobody sane will set such a high premium, but I'm uncomfortable with the panic here. I suggest this instead, which saturates to i64::MAX:

	```rust
	effective_premium.atto().to_i64().unwrap_or(i64::MAX);
	```

**Comment by cryptoAtwill** [comment]
@raulk Updated, please help take another look.

File: fendermint/vm/interpreter/src/fvm/state/priority.rs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+use fvm_shared::bigint::BigInt;
+use fvm_shared::econ::TokenAmount;
+use num_traits::{ToPrimitive, Zero};
+
+/// The transaction priority calculator. The priority calculated is used to determine the ordering
+/// in the mempool.
+#[derive(Clone, Debug)]
+pub struct TxnPriorityCalculator {
+    /// Ring buffer of base fee history
+    base_fee_history: Vec<Option<TokenAmount>>,
```
[the remark this answers is not in the document]
Comment (cryptoAtwill) [reply fendermint/vm/interpreter/src/fvm/state/priority.rs:13]: I think this was quite early discussion on one of the huddles to use the lowest base fee in recent history to let slightly more transaction to go through to avoid empty blocks. An array is also tuneable in how base fee can be selected. The requirement is updated slightly and base fee selection is now includable/non-includable. Then the most recent base fee should be enough. I can update this accordingly.

[the remark this answers is not in the document]
Comment (raulk) [reply fendermint/vm/interpreter/src/fvm/state/priority.rs:13]: We'll need to fix this. The base fee determines inclusion and the burn rate (network fee). By no means should a transaction be included if its gas fee cap is below the current base fee for the block its included in. That's a consensus fault.

**Comment by cryptoAtwill** [comment]
@raulk updated accordingly. Now the priority calculation is based on the latest base fee from the block gas market reading. Anything below will have priority i64::MIN.

File: fendermint/vm/interpreter/src/fvm/state/priority.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+        let msg = create_msg(TokenAmount::from_atto(10), TokenAmount::from_atto(20));
+        assert_eq!(cal.priority(&msg), i64::MIN);
+
+        let msg = create_msg(TokenAmount::from_atto(35), TokenAmount::from_atto(20));
+        assert_eq!(cal.priority(&msg), 5);
+
+        let msg = create_msg(TokenAmount::from_atto(50), TokenAmount::from_atto(20));
+        assert_eq!(cal.priority(&msg), 20);
+
```
Comment (raulk) [inline fendermint/vm/interpreter/src/fvm/state/priority.rs:70]: nit: missing test case for when base fee is below gas fee cap (i.e. includable), and the _effective_ premium is below the indicated premium.

```suggestion

let msg = create_msg(TokenAmount::from_atto(60), TokenAmount::from_atto(20));
        assert_eq!(cal.priority(&msg), 20);
```

**Review by raulk — APPROVED** [review]

**Review by raulk — APPROVED** [review]
An integration test here would be great.

1. Submit many transactions with high gas limit to stuff blocks and make the base fee spike to X or above.
2. Submit a transaction with gas fee cap way below X but above the minimum base fee.
3. Stop the flood so the base fee decreases, and ensure the transaction does not appear until the base fee is below the gas fee cap.

I will merge this and open an issue for these extra tests.

**Comment by raulk** [comment]
@sanderpick feel free to integrate this fix on your side. There are still some edge cases for us to deal with (e.g. rejecting transactions below minimum base fee).

**Comment by sanderpick** [comment]
> @sanderpick feel free to integrate this fix on your side. There are still some edge cases for us to deal with (e.g. rejecting transactions below minimum base fee).

great, ty!
